[17.0][MIG] delivery_purchase: Migration to 17.0

CarlosRoca13 opened the pull request:
cc @Tecnativa TT52416

This PR supersedes PR https://github.com/OCA/delivery-carrier/pull/879 by addressing the changes requested by @rousseldenis and incorporating the changes introduced in PR https://github.com/OCA/delivery-carrier/pull/1024.

ping @pedrobaeza @victoralmau

Commit:
[ADD] delivery_purchase: New module

Co-authored-by: Pedro M. Baeza <pedro.baeza@gmail.com>
Co-authored-by: cubells <vicent@vcubells.net>

[UPD] Update delivery_purchase.pot

victoralmau committed:
[IMP] delivery_purchase: black, isort, prettier

victoralmau committed:
[MIG] delivery_purchase: Migration to 15.0

TT41587

victoralmau committed:
[IMP] delivery_purchase: Add delivery purchase order lines from picking

Auto-create (delivery) order lines when picking are marked as done

TT45656

victoralmau committed:
[FIX] delivery_purchase: Change the position of the carrier to the purchase order header

In v13 it already appeared in the header, and in v14 by mistake it had been moved
to the "Other information" tab because that was where the company_id field was
currently located.
The carrier price is now hidden if there is no carrier set (to avoid confusion).

TT45656

victoralmau committed:
[IMP] delivery_purchase: Set carrier_price when creating a picking

If we have a delivery price of 10 on the order, it is expected to show
10 on the picking (without needing to be set to that amount when set the
picking as done).

TT45656

victoralmau committed:
[IMP] delivery_purchase: Create TestDeliveryPurchaseBase to be used in tests of other modules

TT45656

carolinafernandez-tecnativa committed:
[MIG] delivery_purchase: Migration to 16.0

TT46563

Co-authored-by: carolinafernandez-tecnativa

mymage committed:
Added translation using Weblate (Italian)

victoralmau committed:
[IMP] delivery_purchase: Add Original picking (delivery) (delivery_picking_orig_id) field to purchase order lines

If there are several pickings for an order and several have carriers, several purchase order lines (is_delivery)
will be created but we need to know the origin (picking) that created them.

TT50146

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (14 of 14 strings)

Translation: delivery-carrier-16.0/delivery-carrier-16.0-delivery_purchase
Translate-URL: https://translation.odoo-community.org/projects/delivery-carrier-16-0/delivery-carrier-16-0-delivery_purchase/it/

Commit:
[FIX] delivery_purchase: expected singleton

CarlosRoca13 committed:
[IMP] delivery_purchase: pre-commit auto fixes

rousseldenis commented:
@CarlosRoca13 thanks for this. When supersedes, if possible, you should keep the existing work that has been done by keeping original commits. Thanks

Commit:
[MIG] delivery_purchase: Migration to 17.0

CarlosRoca13 committed:
[IMP] delivery_purchase: Take into account the delivered quantity to calculate the shipping cost

Before these changes, the shipping cost set on the delivery note always took into account the quantity from the purchase order.

After making these changes, the behavior can be modified so that if the system parameter "delivery_purchase.use_delivered_qty_to_set_cost" is set to True, the received quantity will be taken into account when receiving partial quantities on the delivery note.

CarlosRoca13 commented:
> @CarlosRoca13 thanks for this. When supersedes, if possible, you should keep the existing work that has been done by keeping original commits. Thanks

You're right, I attributed the authorship of my first commit to the author of the original PR.

pedrobaeza commented:
/ocabot migration delivery_purchase

pedrobaeza reviewed (commented):
Tested on runboat + code review.

Several comments to improve:

- Please put the commit for #1024 before the migration ones.
- The delivery method and carrier price can be changed after confirming the PO, which doesn't seem correct, specially as this is not propagated to open pickings.
- A PO line is always written when validating the picking. This behavior may be not wanted for some users. Can this be optional by a parameter (also for 15)?

CarlosRoca13 committed:
[IMP] delivery_purchase: Add parameter to avoid the purchase line creation when validate the picking

CarlosRoca13 commented:
Done @pedrobaeza

pedrobaeza commented:
/ocabot merge nobump
